Update `NotificationAssertions`'s  `assert_notifcation` to match against payload subsets and return matched notification (#54126)
rails/rails: activesupport/CHANGELOG.md

    *Jonathan del Strother*

*   `ActiveSupport::Testing::NotificationAssertions`'s `assert_notification` now matches against payload subsets by default.

    Previously the following assertion would fail due to excess key vals in the notification payload. Now with payload subset matching, it will pass.

    ```ruby
    assert_notification("post.submitted", title: "Cool Post") do
      ActiveSupport::Notifications.instrument("post.submitted", title: "Cool Post", body: "Cool Body")
    end
    ```

>     Additionally, one can now persist a matched notification for further, custom assertions.

Reviewer: Updated to match Rails voice more closely. The former copy reads too formal.

```suggestion
    Additionally, you can now persist a matched notification for more customized assertions.
```

    ```ruby
    notification = assert_notification("post.submitted", title: "Cool Post") do
      ActiveSupport::Notifications.instrument("post.submitted", title: "Cool Post", body: Body.new("Cool Body"))
    end

    assert_instance_of(Body, notification.payload[:body])
    ```

>     *Nicholas La Roux*

Reviewer: changelog entries need to go at the top of the file.
